### PR TITLE
Running tests in parallel

### PR DESCRIPTION
**Pull request by potiuk**
This is by far the biggest improvements of the test execution time
we can get now when we are using self-hosted runners.

This change drives down the time of executing all tests on
self-hosted runners from ~ 50 minutes to ~ 13 minutes due to heavy
parallelisation we can implement for different test types and the
fact that our machines for self-hosted runners are far more
capable - they have more CPU, more memory and the fact that
we are using tmpfs for everything.

This change will also drive the cost of our self-hosted runners
down. Since we have auto-scaling infrastructure we will simply need
the machines to run tests for far shorter time. Since the number
of test jobs we run on those self hosted runners is substantial
(10 jobs), we are going to save ~ 6 build hours per one PR/merged
commit!

This also allows the developers to use the power of their
development machines - when you use
`./scripts/ci/testing/ci_run_airflow_testing.sh` the script
detects how many CPU cores are available and it will run as
many parallel test types as many cores you have.

Also in case of Integration tests - they require more memory to run
all the integrations, so in case there is less than ~ 32 GB of RAM
available to Docker, the integration tests are run sequentially
at the end. This drives stability up for machines with lower memory.

On one personal PC (64GB RAM, 8 CPUS/16 cores, fast SSD) the full
test suite execution went down from 30 minutes to 5 minutes.

There is a continuous progress information printed every 10 seconds when
either parallel or sequential tests are run, and the full output is
shown at the end - failed tests are marked in red groups, and succesful are
marked in green groups. This makes it easier to see and analyse errors.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
With LAST rebase I simply rebased a fixup rather than main change :facepalm: . This is the actual change with parallel tests @ashb @kaxil @mik-laj

**Comment by potiuk** [comment]
:pray: approvals please :)

File: scripts/ci/testing/ci_run_airflow_testing.sh [hunk 1]
```diff
@@ -15,77 +15,34 @@
 # KIND, either express or implied.  See the License for the
 # specific language governing permissions and limitations
 # under the License.
+
+# Enable automated tests execution
+RUN_TESTS="true"
+RUN_TESTS="true"
```
Comment (mik-laj) [inline scripts/ci/testing/ci_run_airflow_testing.sh]:
```suggestion
RUN_TESTS="true"
```

File: scripts/ci/testing/ci_run_airflow_testing.sh [hunk 2]
```diff
@@ -109,66 +66,265 @@ function prepare_tests_to_run() {
...
+    echo "${COLOR_BLUE}Print available memory${COLOR_RESET}"
+    echo
+    free --human
+}
 
-for TEST_TYPE in ${TEST_TYPES}
-do
-    start_end::group_start "Running tests ${TEST_TYPE}"
+function get_available_memory_in_docker() {
+    MEMORY_AVAILABLE_FOR_DOCKER=$(docker run --entrypoint /bin/bash \
```
Comment (mik-laj) [inline scripts/ci/testing/ci_run_airflow_testing.sh]:
```suggestion
    MEMORY_AVAILABLE_FOR_DOCKER=$(docker run --rm --entrypoint /bin/bash \
```

File: scripts/ci/testing/ci_run_airflow_testing.sh [hunk 3]
```diff
@@ -109,66 +66,265 @@ function prepare_tests_to_run() {
...
+}
 
-    INTEGRATIONS=()
-    export INTEGRATIONS
 
-    if [[ ${TEST_TYPE:=} == "Integration" ]]; then
-        export ENABLED_INTEGRATIONS="${AVAILABLE_INTEGRATIONS}"
-        export RUN_INTEGRATION_TESTS="${AVAILABLE_INTEGRATIONS}"
+function get_available_cpus_in_docker() {
+    CPUS_AVAILABLE_FOR_DOCKER=$(docker run --entrypoint /bin/bash \
```
Comment (mik-laj) [inline scripts/ci/testing/ci_run_airflow_testing.sh]:
```suggestion
    CPUS_AVAILABLE_FOR_DOCKER=$(docker run --rm --entrypoint /bin/bash \
```

File: scripts/ci/testing/ci_run_airflow_testing.sh [hunk 4]
```diff
@@ -109,66 +66,265 @@ function prepare_tests_to_run() {
...
+    mkdir -p "${PARALLEL_MONITORED_DIR}/${SEMAPHORE_NAME}"
+    for TEST_TYPE in ${test_types_to_run}
+    do
+        export TEST_TYPE
+        mkdir -p "${PARALLEL_MONITORED_DIR}/${SEMAPHORE_NAME}/${TEST_TYPE}"
+        mkdir -p "${PARALLEL_MONITORED_DIR}/${SEMAPHORE_NAME}/${TEST_TYPE}"
+        export JOB_LOG="${PARALLEL_MONITORED_DIR}/${SEMAPHORE_NAME}/${TEST_TYPE}/stdout"
+        export PARALLEL_JOB_STATUS="${PARALLEL_MONITORED_DIR}/${SEMAPHORE_NAME}/${TEST_TYPE}/status"
+        # shellcheck disable=SC2086
+        parallel --ungroup --bg --semaphore --semaphorename "${SEMAPHORE_NAME}" \
```
Comment (mik-laj) [inline scripts/ci/testing/ci_run_airflow_testing.sh:161]: Have you tested it on macOS? I am afraid that this command may not have all the necessary options on this system.

File: scripts/ci/testing/ci_run_airflow_testing.sh (hunk 2 above)
Comment (potiuk) [reply scripts/ci/testing/ci_run_airflow_testing.sh]: Ah yeah. I added it already in #14908 but yeah. I can add it here

File: scripts/ci/testing/ci_run_airflow_testing.sh (hunk 3 above)
Comment (potiuk) [reply scripts/ci/testing/ci_run_airflow_testing.sh]: same

**Review by mik-laj — APPROVED** [review]

File: scripts/ci/testing/ci_run_airflow_testing.sh (hunk 4 above)
Comment (potiuk) [reply scripts/ci/testing/ci_run_airflow_testing.sh:161]: There is no built-in parallel command on MacOS. This script is mainly for CI and having it run on MacOS is a nice-to-have (though useful). 

However I have taken care of it  and tested parallels coming from brew on MacOS (In the scripts I made sure to install GNU Parallel from brew). do not use MacOS with Airflow - but if you do and you could double-check if this command works for MacOS, I would really appreciate.

```
function parallel::make_sure_gnu_parallel_is_installed() {
    start_end::group_start "Making sure GNU Parallels is installed"
    echo
    echo "Making sure you have GNU parallel installed"
    echo
    echo "You might need to provide root password if you do not have it"
    echo
    (command -v parallel || apt install parallel || sudo apt install parallel || brew install parallel) >/dev/null
    start_end::group_end "Making sure GNU Parallels is installed"
}
```

Comment (potiuk) [reply scripts/ci/testing/ci_run_airflow_testing.sh:161]: Ok. I decided to take My Mac  to a spin and it was all surprisingly easy and straightforward. When  I gave my 64GB Mac - 40GB Memory + 8 CPUs it took 581 second (10 minutes) to run full suite of tests (!)

Parallel was not a problem, but I had to remove old commands where I checked memory and disk on host (which is not needed). I remove those and the script works flawlessly on Mac as well.

**Review by kaxil — APPROVED** [review]
Wohoo, let's go

**Comment by potiuk** [comment]
One more thing to check. In non-self-hosted runs I have consistent Exit 137 with MySQL and Python 3.6 when they run 2 test streams in parallel (and I think it's only this combination that has problem in both MySQL versions) . I want to take a closer look at the memory usage pattern for those cases. Maybe it's something I will be able to fix as well.

**Comment by potiuk** [comment]
Monitoring helped a lot:

* I already decreased the use of MySQL memory from ~ 450 MB in peak to 200MB. The default MySQL configuration we used was not optimised at all for our test usage
* However it seems that we have a serious problem in our CLI Tests that often spins out of control:

https://github.com/potiuk/airflow/runs/2158887308?check_suite_focus=true#step:6:4589

Usually our test `airflow` container max at around 400MB, but in cases where we have Exit 137 and failed tests, those spin out of control a lot a

```
  Overall resource statistics
  
  CONTAINER ID                                                       NAME                                         CPU %     MEM USAGE / LIMIT     MEM %     NET I/O           BLOCK I/O        PIDS
  7a4dffd2e4123a8f567704e22d13a41010c3e3f56f6759eac0c56dbb02138bea   airflow-providers_airflow_run_92452ab2adba   22.43%    643MiB / 6.791GiB     9.25%     2.21MB / 2.66MB   21.7MB / 307kB   8
  7605d1260aea0e05c3b5c927a43f85bfd0c89f9165e59218d9b7b5773caf84c1   airflow-providers_mysql_1                    0.20%     94.01MiB / 6.791GiB   1.35%     2.66MB / 2.21MB   36.9kB / 432MB   32
  88d5fd7c47ef1dc1c1269a488b2800183c262a68a3b73a6333bebff2b365e288   airflow-cli_airflow_run_17f7dab0ae27         142.86%   1.829GiB / 6.791GiB   26.93%    9.15MB / 8.32MB   15.1MB / 479kB   54
  c1141763d37ed964f6d6f76e6f842c45e0bc99d5d91d323b3bd069b456c6511d   airflow-cli_mysql_1                          0.19%     104.8MiB / 6.791GiB   1.51%     8.32MB / 9.14MB   684kB / 467MB    38
````

See 1.829GB use of CLI airflow container

and then 

```
  CONTAINER ID                                                       NAME                                         CPU %     MEM USAGE / LIMIT     MEM %     NET I/O           BLOCK I/O        PIDS
  7a4dffd2e4123a8f567704e22d13a41010c3e3f56f6759eac0c56dbb02138bea   airflow-providers_airflow_run_92452ab2adba   58.75%    633.5MiB / 6.791GiB   9.11%     2.26MB / 2.71MB   31.6MB / 307kB   8
  7605d1260aea0e05c3b5c927a43f85bfd0c89f9165e59218d9b7b5773caf84c1   airflow-providers_mysql_1                    4.69%     90MiB / 6.791GiB      1.29%     2.71MB / 2.26MB   1.26MB / 432MB   32
  88d5fd7c47ef1dc1c1269a488b2800183c262a68a3b73a6333bebff2b365e288   airflow-cli_airflow_run_17f7dab0ae27         71.13%    3.781GiB / 6.791GiB   55.68%    9.32MB / 8.43MB   16.1MB / 483kB   42
  c1141763d37ed964f6d6f76e6f842c45e0bc99d5d91d323b3bd069b456c6511d   airflow-cli_mysql_1                          0.05%     101.4MiB / 6.791GiB   1.46%     8.43MB / 9.32MB   684kB / 469MB    38
```

Then it gets to 3.78 GB

This is always around this state of CLI tests:

```
  ### The last lines for CLI process ###
  
  tests/cli/commands/test_sync_perm_command.py .                           [ 67%]
  tests/cli/commands/test_task_command.py ..........s...........
```

so I think it will be rather easy to find the culprit.

**Comment by potiuk** [comment]
Looks like the culprit is `tests/cli/commands/test_task_command.py::TestLogsfromTaskRunCommand::test_logging_with_run_task_subprocess` -> it runs with CAN_FORK = `False` and well, it starts the whole new airlfow process to run the single task and apparently it gets memory up. 

Do we really want to to test it :)  ?

For now  I added additional check for memory available and - similarly to Integration test type I run it sequentially not in parallel

**Comment by potiuk** [comment]
Aaand merged . I quarantined the "Log tests" for now and Added #14937  to fix it in the future - but it seems it's much more stable now.

**Comment by potiuk** [comment]
Let's observe it for a while ...

**Pull request merged**